---
title: "Class 14: Genome Informatics"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Analyze RNA-seq by genotype results

Here we read the results of RNA-seq for about ~230 samples (people) with different genotypes for the rs8067378 Asthma associated SNP.


```{r}
expr <- read.table("rs8067378.txt")
```


```{r}
head(expr)
```


Extract the A/A rows, i.e. where the $geno is equal to "A/A".

```{r}
inds <- expr$geno == "A/A"
expr[inds, ]
```


Print out just the expression values for A/A

```{r}
inds <- expr$geno == "A/A"
expr[inds,]$exp
```

Summary of results for A/A genotype

```{r}
inds <- expr$geno == "A/A"
summary(expr[inds, ]$exp)
```

The summary of the results just G/G

```{r}
inds <- expr$geno == "G/G"
summary(expr[inds, ]$exp)
```


```{r}
boxplot(exp ~ geno, data=expr, notch = TRUE)
```

If notch doesn't overlap, then usually this is statistically significant values.


